tools/lsp/preview/native.rs

```diff
@@ -209,30 +209,30 @@ pub(super) fn open_ui_impl(preview_state: &mut PreviewState) -> Result<(), slint
         }
     };
 
-    super::CONTENT_CACHE.get_or_init(Default::default).lock().unwrap().ui_is_visible = true;
+    preview_state.ui_is_visible = true;
 
     let api = ui.global::<crate::preview::ui::Api>();
     api.set_show_preview_ui(show_preview_ui);
     ui.window().set_fullscreen(fullscreen);
     ui.window().on_close_requested(|| {
-        let mut cache = super::CONTENT_CACHE.get_or_init(Default::default).lock().unwrap();
-        cache.ui_is_visible = false;
+        super::PREVIEW_STATE.with(|preview_state| {
+            let mut preview_state = preview_state.borrow_mut();
+            preview_state.ui_is_visible = false;
+        });
         slint::CloseRequestResponse::HideWindow
     });
     Ok(())
 }
 
+/// Potentially called from other thread!
```

Comment:
i thought you refactor the code so that it is not the case.

But indeed, muda can call it from a different thread
https://github.com/slint-ui/slint/blob/c18b3cf02fc6a8d45e7ec8d2f6f2e665fd429f2b/tools/lsp/preview/native.rs#L361

But then this function won't work if called from a different thread.

Reply:
This entire setup is just too complex. Somebody should do something to get rid of all the complexity of running lsp and live preview in one process and end this madness ;-)

Reply:
This was not resolved.  There is still a call to close_ui from the muda thread
https://github.com/slint-ui/slint/blob/7c1cefd80683b049ad4a2243d053fed7791618fe/tools/lsp/preview/native.rs#L361
and it needs to be wrapped in invoke_from_event_loop

And then you can remove that comment.